### PR TITLE
Enable Electron to set/unset itself as default protocol handler (OS X, Windows)

### LINKED ISSUE #4857: API: Register Electron App as Protocol Handler
There's no simple, cross-platform way to register protocol handlers in Electron. Every OS requires a different solution. This seems like something that Electron should smooth over.

There's already a W3C API for this: [`Navigator.registerProtocolHandler`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler). Can Electron implement it?

### PR DESCRIPTION
**Pull request by felixrieseberg**
This PR adds an API enabling Electron to set itself as the default protocol handler for a custom protocol on both OS X and Windows. The API uses the Windows Registry and LSSetDefaultHandlerForURLScheme internally.

### `app.setAsDefaultProtocolClient(protocol)`

 * `protocol` String - The name of your protocol, without `://`. If you want your app to handle `electron://` links, call this method with `electron` as the parameter.
   
This method sets the current executable as the default handler for a protocol (aka URI scheme). It allows you to integrate your app deeper into the operating system. Once registered, all links with `your-protocol://` will be openend with the current executable. The whole link, including protocol, will be passed to your application as a parameter.
 
On OS X, one can only register protocols that have been added to the app's `info.plist`, which can not be modified at runtime (and is therefore not part of the API). You can however change the file with a simple text editor or script during build time.

### `app.removeAsDefaultProtocolClient(protocol)` _Windows_

  * `protocol` String - The name of your protocol, without `://`.
    
This method checks if the current executable as the default handler for a protocol (aka URI scheme). If so, it will remove the app as the default handler. On OS X, removing the app will automatically remove the app as the default protocol handler.

Closes #4857 || cc @paulcbetts @zcbenz @kevinsawicki

### CODE REVIEW

File: docs/api/app.md [hunk 1]
```diff
@@ -348,6 +348,25 @@ bar, and on OS X you can visit it from dock menu.
 
 Clears the recent documents list.
 
+### `app.setAsDefaultProtocolClient(protocol)` _OS X_ _Windows_
```
Comment (paulcbetts) [inline docs/api/app.md]: How will people unregister from this?

Comment (felixrieseberg) [inline docs/api/app.md]: On both systems, a different app can always overwrite. On OS X, simply removing would involve a bit of hackery (we could call the set with an invalid bundle is); on Windows, we could also delete or overwrite the key. Basically the same it works with magnet links, phone links, or other non-restricted uri schemes.

Both systems don't really have a way to simply "unset" custom uri schemes (in a way that would restore previous handlers, for instance), so I'm inclined to not offer an API that makes Electron users *think* that we can really offer that reliably.

I trust your opinion though, I could easily implement an "overwrite with nonsense" API.

Comment (paulcbetts) [inline docs/api/app.md]: I'm mostly thinking about the uninstall case - on Mac people will delete the app and launchd will react appropriately, but on Windows you leave a trash regkey around

Comment (felixrieseberg) [inline docs/api/app.md]: On the other side, if you check your registry right now, you'll see that we (Microsoft) leave all uri schemes that we ever came across there. We really like to collect them ;-)

Buuuut for a @paulcbetts :+1:, I'd totally throw in a `removeAsDefaultProtocolHandler`, that checks if the current executable is set as handler for a given protocol and removes it. Thoughts?

Comment (paulcbetts) [inline docs/api/app.md]: @felixrieseberg Perfect, works for me

**Comment by deepak1556** [comment]
wouldn't `platform_util` be a better place for the implementation details to go in ?

File: atom/browser/browser_win.cc [hunk 2]
```diff
@@ -125,6 +126,59 @@ void Browser::SetUserTasks(const std::vector<UserTask>& tasks) {
...
+  base::win::RegKey key(root, keyPath.c_str(), KEY_ALL_ACCESS);
+  if (FAILED(key.WriteValue(L"URL Protocol", L"")) ||
+      FAILED(key.WriteValue(L"", urlDecl.c_str())))
+    return false;
+
+  base::win::RegKey commandKey(root, cmdPath.c_str(), KEY_ALL_ACCESS);
+  if (FAILED(commandKey.WriteValue(L"", exe.c_str())))
+    return false;
+
+  VLOG(1) << "Chrome registered as default handler for " << protocol << ".";
```
Comment (deepak1556) [inline atom/browser/browser_win.cc]: log can be removed, or should be modified with applications' name.

**Comment by felixrieseberg** [comment]
Added a `remove` API to round it off.

File: atom/browser/browser_win.cc (hunk 2 above)
Comment (felixrieseberg) [inline atom/browser/browser_win.cc]: Good catch, thanks!

**Comment by deepak1556** [comment]
> given that other APIs that are similar also live in browser.cc - see Dock interactions or executable operations

Yup but the browser class says `// This class is used for control application-wide operations`, so wanted to clarify what should go under `platform_util` and what under `browser`.

File: atom/browser/api/atom_api_app.cc [hunk 3]
```diff
@@ -380,6 +382,8 @@ mate::ObjectTemplateBuilder App::GetObjectTemplateBuilder(
       .SetMethod("setUserTasks",
                  base::Bind(&Browser::SetUserTasks, browser))
       .SetMethod("isAeroGlassEnabled", &App::IsAeroGlassEnabled)
+      .SetMethod("removeAsDefaultProtocolClient",
```
Comment (zcbenz) [inline atom/browser/api/atom_api_app.cc]: Should we provide this API on other platforms even if it does nothing? It is a bit weird that all platforms have `set` but only one platform has `remove`.

**Comment by zcbenz** [comment]
> wouldn't platform_util be a better place for the implementation details to go in ?

The APIs are part of the `app` module, so developers would most likely go to `browser.cc` when they want to look for the code.

**Comment by felixrieseberg** [comment]
@zcbenz: Good point. I added the API to both OS X and Linux, where it just returns `false` (as there's nothing we can do), but it does make it possible for people to call the API on any platform.

I incorporated all feedback, this *should* now be good :+1:

**Comment by zcbenz** [comment]
:+1:

**Comment by atypicalprogrammer** [comment]
Why do you say there is nothing you can do for LInux? 

https://support.shotgunsoftware.com/hc/en-us/articles/219031308-How-to-launch-external-applications-using-custom-protocols-rock-instead-of-http-

http://superuser.com/questions/162092/how-can-i-register-a-custom-protocol-with-xdg

http://www.last.fm/forum/21714/_/42837/1#f430232

http://people.w3.org/~dom/archives/2005/09/integrating-a-new-uris-scheme-handler-to-gnome-and-firefox/

etc..

**Comment by baconface** [comment]
> Is there really nothing that can be done for Linux users? 

Because there isn't a uniform way to do it across distributions.

**Comment by IOTAIedger** [comment]
Some distributions is better than nothing.

**Comment by atypicalprogrammer** [comment]
Some (most popular) distributions is better than nothing.

**Comment by baconface** [comment]
> Some (most popular) distributions is better than nothing.

Unfortunately targeting distributions will be a shot in the dark. As some of the links you pointed out the configuration is done via the window manager or desktop environment. So it's not performed by the OS itself but by Gnome, KDE, XFCE, xdg-open, or some other shell or program. It isn't always promised a distribution will have either of the necessary programs to do this. Even if they do have a shell such a Gnome installed it could still be possible they don't have gnome-open available. Even if you know the distro is ships with them it is possible the user can remove them.

**Comment by baconface** [comment]
I suppose checking for gnome-open, exo-open, xdg-open, gvfs-open, or kde-open and using which ever is available might be better than nothing though. You wouldn't target a distro but you could target possible common environments.

**Pull request merged**